In [22]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.graph_objects as go
import nomarlizer as no
import statsmodels.api as sm
import warnings
import cv2
import fnmatch
import os
import plotly.graph_objects as go
warnings.filterwarnings('ignore')

In [12]:
from load_data import feature_extractor
from scipy import stats
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from skimage.feature import greycomatrix, greycoprops

In [3]:
from outlier import Outlier as ol
from t_test_2sample import ttest_2samp as t_test

In [4]:
def tester(pred, k):
    count = 0
    for i in pred:
        if i == k:
            count += 1

    print(f'accuracy: {count/len(pred)*100:.2f}')

In [5]:
def anova_test(df1, df2):
    for x in df1.columns:
        f_value, p_value = stats.f_oneway(df1[x], df2[x])
        print(f'{x} = f_value: {f_value:.4f}, p_value: {p_value:.4f}')


In [6]:
# roi = [(5, 0), (6, 0), (7, 0), (5, 1), (6, 1), (7, 1)]   # weather_roi
roi = [(1, 14), (1, 15), (2, 14), (2, 15), (3, 14), (3, 15)]   # road_roi

In [31]:
columns = ['h_1', 'h_2', 'h_3', 'h_4', 'h_5', 'h_6', 'h_7', 'h_8', 'h_9', 'h_10', 'h_11', 'h_12', 'h_13', 'lbp_1', 'lbp_2', 'lbp_3', 'lbp_4', 'lbp_5', 'lbp_6', 'lbp_7', 'lbp_8', 'lbp_9', 'lbp_10', 'bright']

In [9]:
source_path = r'D:/analysis/dry_day_201'

In [19]:
source_path = source_path.replace('\\', '/')
height, width = 150, 150
file_list = fnmatch.filter(os.listdir(source_path), '*.jpg')

glcm_list = []
xs, ys = [], []

for img in file_list:
    path = os.path.join(source_path, img)
    img = cv2.imread(path)
    for (i, j) in roi:
        x, y = i * width, j * height
        roi_img = img[y:y + width, x:x + height]
        roi_img = cv2.cvtColor(roi_img, cv2.COLOR_BGR2GRAY)
        glcm = greycomatrix(roi_img, distances=[5], angles=[0], levels=256, symmetric=True, normed=True)
        glcm_list.append(glcm)
        xs.append(greycoprops(glcm, 'dissimilarity')[0, 0])
        ys.append(greycoprops(glcm, 'correlation')[0, 0])

In [24]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=xs, y=ys, mode='markers'))

In [26]:
source_path = r'D:/analysis/wet_day_201'

In [27]:
source_path = source_path.replace('\\', '/')
height, width = 150, 150
file_list = fnmatch.filter(os.listdir(source_path), '*.jpg')

glcm_list = []
xs_r, ys_r = [], []

for img in file_list:
    path = os.path.join(source_path, img)

    img = cv2.imread(path)

    for (i, j) in roi:
        x, y = i * width, j * height
        roi_img = img[y:y + width, x:x + height]
        roi_img = cv2.cvtColor(roi_img, cv2.COLOR_BGR2GRAY)
        glcm = greycomatrix(roi_img, distances=[5], angles=[0], levels=256, symmetric=True, normed=True)
        glcm_list.append(glcm)
        xs_r.append(greycoprops(glcm, 'dissimilarity')[0, 0])
        ys_r.append(greycoprops(glcm, 'correlation')[0, 0])

In [30]:
source_path = r'D:/analysis/dry_day_203'
source_path = source_path.replace('\\', '/')
height, width = 150, 150
file_list = fnmatch.filter(os.listdir(source_path), '*.jpg')

glcm_list = []
xs_2, ys_2 = [], []

for img in file_list:
    path = os.path.join(source_path, img)

    img = cv2.imread(path)

    for (i, j) in roi:
        x, y = i * width, j * height
        roi_img = img[y:y + width, x:x + height]
        roi_img = cv2.cvtColor(roi_img, cv2.COLOR_BGR2GRAY)
        glcm = greycomatrix(roi_img, distances=[5], angles=[0], levels=256, symmetric=True, normed=True)
        glcm_list.append(glcm)
        xs_2.append(greycoprops(glcm, 'dissimilarity')[0, 0])
        ys_2.append(greycoprops(glcm, 'correlation')[0, 0])

In [31]:
source_path = r'D:/analysis/wet_day_203'
source_path = source_path.replace('\\', '/')
height, width = 150, 150
file_list = fnmatch.filter(os.listdir(source_path), '*.jpg')

glcm_list = []
xs_r2, ys_r2 = [], []

for img in file_list:
    path = os.path.join(source_path, img)

    img = cv2.imread(path)

    for (i, j) in roi:
        x, y = i * width, j * height
        roi_img = img[y:y + width, x:x + height]
        roi_img = cv2.cvtColor(roi_img, cv2.COLOR_BGR2GRAY)
        glcm = greycomatrix(roi_img, distances=[5], angles=[0], levels=256, symmetric=True, normed=True)
        glcm_list.append(glcm)
        xs_r2.append(greycoprops(glcm, 'dissimilarity')[0, 0])
        ys_r2.append(greycoprops(glcm, 'correlation')[0, 0])

In [32]:
fig = go.Figure()
fig.add_trace(go.Scatter(x=xs, y=ys, mode='markers', name='dd_201'))
fig.add_trace(go.Scatter(x=xs_r, y=ys_r, mode='markers', name='wd_201'))
fig.add_trace(go.Scatter(x=xs_2, y=ys_2, mode='markers', name='dd_203'))
fig.add_trace(go.Scatter(x=xs_r2, y=ys_r2, mode='markers', name='wd_203'))
fig.show()

In [32]:
# Data load

dd_test_201 = feature_extractor('D:/analysis/dry_day_201', roi)
dd_test_203 = feature_extractor('D:/analysis/dry_day_203', roi)

wd_test_201 = feature_extractor('D:/analysis/wet_day_201', roi)
wd_test_203 = feature_extractor('D:/analysis/wet_day_203', roi)

In [7]:
def get_pixel(img, center, x, y): 
      
    new_value = 0
      
    try: 
        # If local neighbourhood pixel  
        # value is greater than or equal 
        # to center pixel values then  
        # set it to 1 
        if img[x][y] >= center: 
            new_value = 1
              
    except: 
        # Exception is required when  
        # neighbourhood value of a center 
        # pixel value is null i.e. values 
        # present at boundaries. 
        pass
      
    return new_value 
   
# Function for calculating LBP 
def lbp_calculated_pixel(img, x, y): 
   
    center = img[x][y] 
   
    val_ar = [] 
      
    # top_left 
    val_ar.append(get_pixel(img, center, x-1, y-1)) 
      
    # top 
    val_ar.append(get_pixel(img, center, x-1, y)) 
      
    # top_right 
    val_ar.append(get_pixel(img, center, x-1, y + 1)) 
      
    # right 
    val_ar.append(get_pixel(img, center, x, y + 1)) 
      
    # bottom_right 
    val_ar.append(get_pixel(img, center, x + 1, y + 1)) 
      
    # bottom 
    val_ar.append(get_pixel(img, center, x + 1, y)) 
      
    # bottom_left 
    val_ar.append(get_pixel(img, center, x + 1, y-1)) 
      
    # left 
    val_ar.append(get_pixel(img, center, x, y-1)) 
       
    # Now, we need to convert binary 
    # values to decimal 
    power_val = [1, 2, 4, 8, 16, 32, 64, 128] 
   
    val = 0
      
    for i in range(len(val_ar)): 
        val += val_ar[i] * power_val[i] 
          
    return val 

In [8]:
img = cv2.imread('D:/road_203/OPTICAL203_20210301121822607175_.jpg', 1)

In [9]:
cv2.imshow('aa', img)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [10]:
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

In [11]:
cv2.imshow('aa', gray)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [13]:
height, width, _ = img.shape

In [14]:
img_lbp = np.zeros((height, width), np.uint8)

In [16]:
for i in range(0, height): 
    for j in range(0, width): 
        img_lbp[i, j] = lbp_calculated_pixel(gray, i, j) 

In [17]:
cv2.imshow('aa', img_lbp)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [18]:
dd_img = cv2.imread('D:/analysis/dry_day_203/OPTICAL203_20210201104630118878_.jpg', 1)

In [19]:
dd_gray = cv2.cvtColor(dd_img, cv2.COLOR_BGR2GRAY)

In [20]:
height, width, _ = dd_img.shape

In [21]:
img_lbp = np.zeros((height, width), np.uint8)

In [22]:
for i in range(0, height): 
    for j in range(0, width): 
        img_lbp[i, j] = lbp_calculated_pixel(dd_gray, i, j) 

In [23]:
cv2.imshow('aa', img_lbp)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [53]:
# create Dataframe
dd1_df = pd.DataFrame(dd_test_201, columns=columns)
dd3_df = pd.DataFrame(dd_test_203, columns=columns)

wd1_df = pd.DataFrame(wd_test_201, columns=columns)
wd3_df = pd.DataFrame(wd_test_203, columns=columns)

In [22]:
# del outlier
dd1_df_ol = ol(dd1_df)
dd3_df_ol = ol(dd3_df)

wd1_df_ol = ol(wd1_df)
wd3_df_ol = ol(wd3_df)

In [23]:
# z_score normalizer
dd1_df_no = no.normalizer(dd1_df)
dd3_df_no = no.normalizer(dd3_df)

wd1_df_no = no.normalizer(wd1_df)
wd3_df_no = no.normalizer(wd3_df)

In [24]:
# del outlier z_score normalizer
dd1_ol_no = no.normalizer(dd1_df_ol)
dd3_ol_no = no.normalizer(dd3_df_ol)

wd1_ol_no = no.normalizer(wd1_df_ol)
wd3_ol_no = no.normalizer(wd3_df_ol)

In [25]:
t_test(dd1_df, dd3_df)

h_1 = t_value: -41.6340, p_value: 0.0000
h_2 = t_value: 30.1601, p_value: 0.0000
h_3 = t_value: -38.7801, p_value: 0.0000
h_4 = t_value: 4.7288, p_value: 0.0000
h_5 = t_value: -58.3500, p_value: 0.0000
h_6 = t_value: 9.1265, p_value: 0.0000
h_7 = t_value: 3.5561, p_value: 0.0004
h_8 = t_value: 25.7713, p_value: 0.0000
h_9 = t_value: 49.1522, p_value: 0.0000
h_10 = t_value: -61.8619, p_value: 0.0000
h_11 = t_value: 55.6498, p_value: 0.0000
h_12 = t_value: 56.1263, p_value: 0.0000
h_13 = t_value: -33.8417, p_value: 0.0000
lbp_1 = t_value: 62.5662, p_value: 0.0000
lbp_2 = t_value: 63.7767, p_value: 0.0000
lbp_3 = t_value: -12.2456, p_value: 0.0000
lbp_4 = t_value: -62.5484, p_value: 0.0000
lbp_5 = t_value: -44.2236, p_value: 0.0000
lbp_6 = t_value: -61.6470, p_value: 0.0000
lbp_7 = t_value: -59.9083, p_value: 0.0000
lbp_8 = t_value: -16.8959, p_value: 0.0000
lbp_9 = t_value: -22.1775, p_value: 0.0000
lbp_10 = t_value: 58.9157, p_value: 0.0000
bright = t_value: 8.9128, p_value: 0.0000
0.05

In [26]:
t_test(dd1_df, wd1_df)

h_1 = t_value: -10.4412, p_value: 0.0000
h_2 = t_value: 23.3916, p_value: 0.0000
h_3 = t_value: 18.1852, p_value: 0.0000
h_4 = t_value: 25.8572, p_value: 0.0000
h_5 = t_value: -5.0070, p_value: 0.0000
h_6 = t_value: 9.7308, p_value: 0.0000
h_7 = t_value: 25.0051, p_value: 0.0000
h_8 = t_value: 48.2277, p_value: 0.0000
h_9 = t_value: 25.8578, p_value: 0.0000
h_10 = t_value: -10.9877, p_value: 0.0000
h_11 = t_value: 14.5845, p_value: 0.0000
h_12 = t_value: -17.1505, p_value: 0.0000
h_13 = t_value: 17.8775, p_value: 0.0000
lbp_1 = t_value: 3.6333, p_value: 0.0003
lbp_2 = t_value: -11.7895, p_value: 0.0000
lbp_3 = t_value: -14.4932, p_value: 0.0000
lbp_4 = t_value: 0.6984, p_value: 0.4850
lbp_5 = t_value: 5.2005, p_value: 0.0000
lbp_6 = t_value: 4.1984, p_value: 0.0000
lbp_7 = t_value: -4.9745, p_value: 0.0000
lbp_8 = t_value: -15.3910, p_value: 0.0000
lbp_9 = t_value: 20.1454, p_value: 0.0000
lbp_10 = t_value: -5.1074, p_value: 0.0000
bright = t_value: 9.8610, p_value: 0.0000
0.05이하 채택 된 

In [27]:
t_test(dd3_df, wd3_df)

h_1 = t_value: -44.8565, p_value: 0.0000
h_2 = t_value: 22.3096, p_value: 0.0000
h_3 = t_value: 15.6319, p_value: 0.0000
h_4 = t_value: 22.4166, p_value: 0.0000
h_5 = t_value: -25.0259, p_value: 0.0000
h_6 = t_value: 15.8019, p_value: 0.0000
h_7 = t_value: 22.3838, p_value: 0.0000
h_8 = t_value: 56.7025, p_value: 0.0000
h_9 = t_value: 47.7644, p_value: 0.0000
h_10 = t_value: -35.5973, p_value: 0.0000
h_11 = t_value: 25.7670, p_value: 0.0000
h_12 = t_value: 2.1837, p_value: 0.0290
h_13 = t_value: 7.1626, p_value: 0.0000
lbp_1 = t_value: 15.9766, p_value: 0.0000
lbp_2 = t_value: 16.7413, p_value: 0.0000
lbp_3 = t_value: 25.2328, p_value: 0.0000
lbp_4 = t_value: 1.4207, p_value: 0.1555
lbp_5 = t_value: 14.3413, p_value: 0.0000
lbp_6 = t_value: -11.9333, p_value: 0.0000
lbp_7 = t_value: -19.9318, p_value: 0.0000
lbp_8 = t_value: -21.3290, p_value: 0.0000
lbp_9 = t_value: -19.5615, p_value: 0.0000
lbp_10 = t_value: 12.6646, p_value: 0.0000
bright = t_value: 16.1131, p_value: 0.0000
0.05이하 채

In [28]:
t_test(dd1_df_ol, dd3_df_ol)

h_1 = t_value: -51.1945, p_value: 0.0000
h_2 = t_value: 35.0023, p_value: 0.0000
h_3 = t_value: -43.8869, p_value: 0.0000
h_4 = t_value: 27.6763, p_value: 0.0000
h_5 = t_value: -58.9513, p_value: 0.0000
h_6 = t_value: 9.0510, p_value: 0.0000
h_7 = t_value: 25.7175, p_value: 0.0000
h_8 = t_value: 32.9804, p_value: 0.0000
h_9 = t_value: 50.0227, p_value: 0.0000
h_10 = t_value: -63.0975, p_value: 0.0000
h_11 = t_value: 55.6498, p_value: 0.0000
h_12 = t_value: 56.7676, p_value: 0.0000
h_13 = t_value: -37.2522, p_value: 0.0000
lbp_1 = t_value: 62.8896, p_value: 0.0000
lbp_2 = t_value: 63.8911, p_value: 0.0000
lbp_3 = t_value: -12.1195, p_value: 0.0000
lbp_4 = t_value: -64.1882, p_value: 0.0000
lbp_5 = t_value: -55.6989, p_value: 0.0000
lbp_6 = t_value: -63.0124, p_value: 0.0000
lbp_7 = t_value: -59.9083, p_value: 0.0000
lbp_8 = t_value: -21.7160, p_value: 0.0000
lbp_9 = t_value: -30.3171, p_value: 0.0000
lbp_10 = t_value: 59.5994, p_value: 0.0000
bright = t_value: 8.3035, p_value: 0.0000
0.

In [29]:
t_test(dd1_df_ol, wd1_df_ol)

h_1 = t_value: -14.1138, p_value: 0.0000
h_2 = t_value: 23.6876, p_value: 0.0000
h_3 = t_value: 18.1852, p_value: 0.0000
h_4 = t_value: 34.6819, p_value: 0.0000
h_5 = t_value: -5.0070, p_value: 0.0000
h_6 = t_value: 13.0482, p_value: 0.0000
h_7 = t_value: 38.0893, p_value: 0.0000
h_8 = t_value: 48.5702, p_value: 0.0000
h_9 = t_value: 25.8578, p_value: 0.0000
h_10 = t_value: -10.6785, p_value: 0.0000
h_11 = t_value: 14.5845, p_value: 0.0000
h_12 = t_value: -18.7921, p_value: 0.0000
h_13 = t_value: 18.5282, p_value: 0.0000
lbp_1 = t_value: 3.6333, p_value: 0.0003
lbp_2 = t_value: -13.4621, p_value: 0.0000
lbp_3 = t_value: -14.4631, p_value: 0.0000
lbp_4 = t_value: 0.3432, p_value: 0.7315
lbp_5 = t_value: 2.7345, p_value: 0.0063
lbp_6 = t_value: 5.8046, p_value: 0.0000
lbp_7 = t_value: -4.9745, p_value: 0.0000
lbp_8 = t_value: -14.8722, p_value: 0.0000
lbp_9 = t_value: 27.6616, p_value: 0.0000
lbp_10 = t_value: -6.5729, p_value: 0.0000
bright = t_value: 13.5219, p_value: 0.0000
0.05이하 채택 

In [30]:
t_test(dd3_df_ol, wd3_df_ol)

h_1 = t_value: -52.5976, p_value: 0.0000
h_2 = t_value: 31.9920, p_value: 0.0000
h_3 = t_value: 29.8396, p_value: 0.0000
h_4 = t_value: 61.7606, p_value: 0.0000
h_5 = t_value: -31.1806, p_value: 0.0000
h_6 = t_value: 23.8148, p_value: 0.0000
h_7 = t_value: 61.5011, p_value: 0.0000
h_8 = t_value: 62.8132, p_value: 0.0000
h_9 = t_value: 52.5786, p_value: 0.0000
h_10 = t_value: -38.0043, p_value: 0.0000
h_11 = t_value: 29.6839, p_value: 0.0000
h_12 = t_value: 3.5145, p_value: 0.0004
h_13 = t_value: 27.8991, p_value: 0.0000
lbp_1 = t_value: 19.9838, p_value: 0.0000
lbp_2 = t_value: 20.8432, p_value: 0.0000
lbp_3 = t_value: 25.6883, p_value: 0.0000
lbp_4 = t_value: -0.3551, p_value: 0.7226
lbp_5 = t_value: 9.7051, p_value: 0.0000
lbp_6 = t_value: -12.8204, p_value: 0.0000
lbp_7 = t_value: -23.5913, p_value: 0.0000
lbp_8 = t_value: -21.5139, p_value: 0.0000
lbp_9 = t_value: -20.5749, p_value: 0.0000
lbp_10 = t_value: 14.2365, p_value: 0.0000
bright = t_value: 26.0845, p_value: 0.0000
0.05이하 

In [31]:
anova_test(dd1_df, wd1_df)

h_1 = f_value: 109.0190, p_value: 0.0000
h_2 = f_value: 547.1691, p_value: 0.0000
h_3 = f_value: 330.7016, p_value: 0.0000
h_4 = f_value: 668.5961, p_value: 0.0000
h_5 = f_value: 25.0704, p_value: 0.0000
h_6 = f_value: 94.6877, p_value: 0.0000
h_7 = f_value: 625.2556, p_value: 0.0000
h_8 = f_value: 2325.9130, p_value: 0.0000
h_9 = f_value: 668.6269, p_value: 0.0000
h_10 = f_value: 120.7287, p_value: 0.0000
h_11 = f_value: 212.7084, p_value: 0.0000
h_12 = f_value: 294.1405, p_value: 0.0000
h_13 = f_value: 319.6059, p_value: 0.0000
lbp_1 = f_value: 13.2007, p_value: 0.0003
lbp_2 = f_value: 138.9912, p_value: 0.0000
lbp_3 = f_value: 210.0542, p_value: 0.0000
lbp_4 = f_value: 0.4877, p_value: 0.4850
lbp_5 = f_value: 27.0451, p_value: 0.0000
lbp_6 = f_value: 17.6270, p_value: 0.0000
lbp_7 = f_value: 24.7458, p_value: 0.0000
lbp_8 = f_value: 236.8832, p_value: 0.0000
lbp_9 = f_value: 405.8380, p_value: 0.0000
lbp_10 = f_value: 26.0857, p_value: 0.0000
bright = f_value: 97.2402, p_value: 0.00

In [32]:
anova_test(dd1_df_ol, wd1_df_ol)

h_1 = f_value: 199.1991, p_value: 0.0000
h_2 = f_value: 561.1030, p_value: 0.0000
h_3 = f_value: 330.7016, p_value: 0.0000
h_4 = f_value: 1202.8356, p_value: 0.0000
h_5 = f_value: 25.0704, p_value: 0.0000
h_6 = f_value: 170.2563, p_value: 0.0000
h_7 = f_value: 1450.7925, p_value: 0.0000
h_8 = f_value: 2359.0647, p_value: 0.0000
h_9 = f_value: 668.6269, p_value: 0.0000
h_10 = f_value: 114.0307, p_value: 0.0000
h_11 = f_value: 212.7084, p_value: 0.0000
h_12 = f_value: 353.1412, p_value: 0.0000
h_13 = f_value: 343.2932, p_value: 0.0000
lbp_1 = f_value: 13.2007, p_value: 0.0003
lbp_2 = f_value: 181.2287, p_value: 0.0000
lbp_3 = f_value: 209.1800, p_value: 0.0000
lbp_4 = f_value: 0.1178, p_value: 0.7315
lbp_5 = f_value: 7.4774, p_value: 0.0063
lbp_6 = f_value: 33.6930, p_value: 0.0000
lbp_7 = f_value: 24.7458, p_value: 0.0000
lbp_8 = f_value: 221.1829, p_value: 0.0000
lbp_9 = f_value: 765.1650, p_value: 0.0000
lbp_10 = f_value: 43.2027, p_value: 0.0000
bright = f_value: 182.8420, p_value: 0

In [130]:
dd1_df['target'] = 0
dd3_df['target'] = 1
wd1_df['target'] = 2
wd3_df['target'] = 3

In [59]:
data_dry = pd.concat([dd1_df, dd3_df], axis=1)
data_wet = pd.concat([wd1_df, wd3_df], axis=1)
data_dry.reset_index(inplace=True, drop=True)
data_wet.reset_index(inplace=True, drop=True)

In [60]:
data_dry['target'] = 0
data_wet['target'] = 1

In [131]:
data = pd.concat([dd1_df, dd3_df, wd1_df, wd3_df], axis=0)
data.reset_index(inplace=True, drop=True)

In [61]:
total_data = pd.concat([data_dry, data_wet], axis=0)
total_data.reset_index(inplace=True, drop=True)

In [63]:
total_data

h_1        h_2       h_3        h_4       h_5         h_6  \
0     0.004249  39.600540  0.616880  51.724330  0.277161  278.715251   
1     0.002704  43.131528  0.772732  94.900740  0.267291  262.505695   
2     0.001660  50.052290  0.717742  88.706920  0.198975  252.671326   
3     0.002780  43.777802  0.585919  52.864535  0.227660  234.432181   
4     0.002051  42.213222  0.701762  70.789492  0.228693  260.102790   
...        ...        ...       ...        ...       ...         ...   
4075  0.017448   6.756051  0.800598  16.940586  0.576759  228.280113   
4076  0.031746   3.107798  0.815274   8.418836  0.634757  241.500692   
4077  0.029947   2.575771  0.847585   8.449651  0.650453  233.940591   
4078  0.027350   1.375624  0.947626  13.132259  0.722587  242.781399   
4079  0.033725   1.470899  0.923100   9.569890  0.722990  229.125034   

             h_7       h_8       h_9      h_10  ...     lbp_3     lbp_4  \
0     167.296781  5.567350  8.895602  0.000626  ...  0.034889  0.073867   
1     336.471433  6.084120  9.507177  0.000632  ...  0.052800  0.075778   
2     304.775388  6.114454  9.811530  0.000514  ...  0.065156  0.074089   
3     167.680337  5.702623  9.265208  0.000598  ...  0.048044  0.082978   
4     240.944748  5.980316  9.520294  0.000572  ...  0.031244  0.074667   
...          ...       ...       ...       ...  ...       ...       ...   
4075   61.006293  4.961762  6.803908  0.001888  ...  0.032444  0.078222   
4076   30.567548  4.442000  5.876937  0.002370  ...  0.023689  0.081867   
4077   31.222831  4.499991  5.881482  0.002601  ...  0.025067  0.073778   
4078   51.153413  4.775089  5.831177  0.002982  ...  0.029467  0.080978   
4079   36.808663  4.594691  5.624009  0.003142  ...  0.037289  0.089156   

         lbp_5     lbp_6     lbp_7     lbp_8     lbp_9    lbp_10  bright  \
0     0.079422  0.153067  0.096756  0.131200  0.204044  0.153733   147.0   
1     0.089600  0.115156  0.087200  0.110311  0.120800  0.232311   141.0   
2     0.084622  0.094000  0.078578  0.107022  0.084578  0.268533   143.0   
3     0.109644  0.144800  0.099644  0.107600  0.101511  0.205333   140.0   
4     0.089822  0.164756  0.103289  0.124222  0.191556  0.152311   139.0   
...        ...       ...       ...       ...       ...       ...     ...   
4075  0.091378  0.173956  0.100267  0.125867  0.200400  0.145067   121.0   
4076  0.092756  0.182578  0.101556  0.125778  0.230178  0.120889   128.0   
4077  0.086133  0.190933  0.111556  0.119244  0.221733  0.124444   121.0   
4078  0.087822  0.171511  0.102356  0.123156  0.210756  0.139822   127.0   
4079  0.101067  0.176489  0.099467  0.112400  0.161422  0.164400   120.0   

      target  
0          0  
1          0  
2          0  
3          0  
4          0  
...      ...  
4075       1  
4076       1  
4077       1  
4078       1  
4079       1  

[4080 rows x 49 columns]

In [132]:
x_train, x_test, y_train, y_test = train_test_split(data.iloc[:, :-1], data.iloc[:, -1], test_size=0.2, random_state=2021)

In [64]:
x_train, x_test, y_train, y_test = train_test_split(total_data.iloc[:, :-1], total_data.iloc[:, -1], test_size=0.2, random_state=2021)

In [218]:
x_total_data = total_data.iloc[:, :-1]
y_total_data = total_data.iloc[:, -1]

In [196]:
x_data = data.iloc[:, :-1]
y_data = data.iloc[:, -1]

In [254]:
params = {
    'C' : [x for x in range(1, 101)]
}

In [255]:
test_svc = SVC(random_state=2021)

In [256]:
grid_svc = GridSearchCV(test_svc, params, verbose=2, cv=3)

In [257]:
grid_svc.fit(x_total_data, y_total_data)

Fitting 3 folds for each of 100 candidates, totalling 300 fits
[CV] END ................................................C=1; total time=   0.1s
[CV] END ................................................C=1; total time=   0.1s
[CV] END ................................................C=1; total time=   0.2s
[CV] END ................................................C=2; total time=   0.1s
[CV] END ................................................C=2; total time=   0.1s
[CV] END ................................................C=2; total time=   0.2s
[CV] END ................................................C=3; total time=   0.1s
[CV] END ................................................C=3; total time=   0.1s
[CV] END ................................................C=3; total time=   0.1s
[CV] END ................................................C=4; total time=   0.1s
[CV] END ................................................C=4; total time=   0.1s
[CV] END .....................................

[CV] END ...............................................C=35; total time=   0.0s
[CV] END ...............................................C=35; total time=   0.0s
[CV] END ...............................................C=35; total time=   0.1s
[CV] END ...............................................C=36; total time=   0.0s
[CV] END ...............................................C=36; total time=   0.0s
[CV] END ...............................................C=36; total time=   0.1s
[CV] END ...............................................C=37; total time=   0.0s
[CV] END ...............................................C=37; total time=   0.0s
[CV] END ...............................................C=37; total time=   0.1s
[CV] END ...............................................C=38; total time=   0.0s
[CV] END ...............................................C=38; total time=   0.0s
[CV] END ...............................................C=38; total time=   0.1s
[CV] END ...................

[CV] END ...............................................C=69; total time=   0.0s
[CV] END ...............................................C=69; total time=   0.0s
[CV] END ...............................................C=69; total time=   0.1s
[CV] END ...............................................C=70; total time=   0.0s
[CV] END ...............................................C=70; total time=   0.0s
[CV] END ...............................................C=70; total time=   0.1s
[CV] END ...............................................C=71; total time=   0.0s
[CV] END ...............................................C=71; total time=   0.0s
[CV] END ...............................................C=71; total time=   0.1s
[CV] END ...............................................C=72; total time=   0.0s
[CV] END ...............................................C=72; total time=   0.0s
[CV] END ...............................................C=72; total time=   0.1s
[CV] END ...................

GridSearchCV(cv=3, estimator=SVC(random_state=2021),
             param_grid={'C': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14,
                               15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
                               27, 28, 29, 30, ...]},
             verbose=2)

In [258]:
grid_svc.best_estimator_

SVC(C=100, random_state=2021)

In [259]:
svc_params = {
    'C': stats.uniform(1, 20),
    'gamma': stats.reciprocal(0.001, 0.1),
    'kernel': ['poly', 'linear', 'rbf'],

}

In [260]:
svc_clf = SVC(random_state=2021, class_weight={0: 0.8, 1:1.5})

In [261]:
svc_rsv = RandomizedSearchCV(estimator=svc_clf, param_distributions=svc_params, n_iter=30, verbose=2, cv=3)

In [ ]:
svc_rsv.fit(x_total_data, y_total_data)

Fitting 3 folds for each of 30 candidates, totalling 90 fits
[CV] END C=1.6459464675467186, gamma=0.04136479090708694, kernel=linear; total time=   2.4s
[CV] END C=1.6459464675467186, gamma=0.04136479090708694, kernel=linear; total time=  10.5s
[CV] END C=1.6459464675467186, gamma=0.04136479090708694, kernel=linear; total time=  25.8s
[CV] END C=18.86407008464891, gamma=0.010999717232785053, kernel=linear; total time=   9.1s
[CV] END C=18.86407008464891, gamma=0.010999717232785053, kernel=linear; total time=  37.0s
[CV] END C=18.86407008464891, gamma=0.010999717232785053, kernel=linear; total time=  53.4s
[CV] END C=19.70908225363722, gamma=0.010285201433325357, kernel=poly; total time= 1.4min
[CV] END C=19.70908225363722, gamma=0.010285201433325357, kernel=poly; total time= 6.8min
[CV] END C=19.70908225363722, gamma=0.010285201433325357, kernel=poly; total time=19.3min
[CV] END C=3.69495338944596, gamma=0.001687832884697058, kernel=poly; total time=  21.5s
[CV] END C=3.69495338944596,

In [ ]:
svc_rsv_2 = RandomizedSearchCV(estimator=svc_clf, param_distributions=svc_params, n_iter=30, verbose=2, cv=3)

In [ ]:
svc_rsv.fit(x_data, y_data)

In [240]:
svc_rsv.best_estimator_

SVC(C=9.426150967189454, class_weight={0: 0.8, 1: 1.5},
    gamma=0.0012734275447072905, random_state=2021)

In [241]:
svc_rsv.best_estimator_.fit(x_total_data, y_total_data)

SVC(C=9.426150967189454, class_weight={0: 0.8, 1: 1.5},
    gamma=0.0012734275447072905, random_state=2021)

In [244]:
svc_pipe = make_pipeline(StandardScaler(),RandomizedSearchCV(estimator=svc_clf, param_distributions=svc_params, n_iter=30, verbose=2, cv=3))

In [245]:
svc_pipe.fit(x_total_data, y_total_data)

Fitting 3 folds for each of 30 candidates, totalling 90 fits
[CV] END ...C=18.95267030505911, gamma=0.0024024141380908233; total time=   0.0s
[CV] END ...C=18.95267030505911, gamma=0.0024024141380908233; total time=   0.0s
[CV] END ...C=18.95267030505911, gamma=0.0024024141380908233; total time=   0.0s
[CV] END .....C=9.018946187733608, gamma=0.02791948063399854; total time=   0.0s
[CV] END .....C=9.018946187733608, gamma=0.02791948063399854; total time=   0.0s
[CV] END .....C=9.018946187733608, gamma=0.02791948063399854; total time=   0.0s
[CV] END ....C=17.06057091411664, gamma=0.005494606281648472; total time=   0.0s
[CV] END ....C=17.06057091411664, gamma=0.005494606281648472; total time=   0.0s
[CV] END ....C=17.06057091411664, gamma=0.005494606281648472; total time=   0.0s
[CV] END ....C=11.966226218401609, gamma=0.01801011838636657; total time=   0.0s
[CV] END ....C=11.966226218401609, gamma=0.01801011838636657; total time=   0.0s
[CV] END ....C=11.966226218401609, gamma=0.01801

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('randomizedsearchcv',
                 RandomizedSearchCV(cv=3,
                                    estimator=SVC(class_weight={0: 0.8, 1: 1.5},
                                                  gamma='auto',
                                                  random_state=2021),
                                    n_iter=30,
                                    param_distributions={'C': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001C8713C72B0>,
                                                         'gamma': <scipy.stats._distn_infrastructure.rv_frozen object at 0x000001C8713C78E0>},
                                    verbose=2))])

In [247]:
svc_pipe.classes_

array([0, 1], dtype=int64)

In [251]:
std = StandardScaler()
test_data = std.fit_transform(test_total_day)

In [252]:
svc_pipe.predict(test_data)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [242]:
pred = svc_rsv.predict(test_total_wet)

In [243]:
tester(pred, 1)

accuracy: 0.33


In [133]:
svc = SVC(random_state=2021)

In [134]:
params = {
    'C' : [x  for x in range(1, 100)],
}

In [135]:
gridsvc = GridSearchCV(estimator=svc, param_grid=params)

In [136]:
gridsvc.fit(x_train, y_train)

GridSearchCV(estimator=SVC(random_state=2021),
             param_grid={'C': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14,
                               15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
                               27, 28, 29, 30, ...]})

In [137]:
gridsvc.best_params_

{'C': 96}

In [107]:
c = gridsvc.best_params_.values()

In [138]:
svc_grid = SVC(C = 96, random_state=2021)

In [139]:
svc_grid.fit(x_train, y_train)

SVC(C=96, random_state=2021)

In [204]:
test_set = r'D:/analysis/201_203/53R_201/dry_day'
test_set2 = r'D:/analysis/201_203/53R_203/dry_day'
test_set3 = r'D:/analysis/201_203/53R_201/wet_day'
test_set4 = r'D:/analysis/201_203/53R_203/wet_day'

In [212]:
test_set = feature_extractor(test_set, roi)
test_set2 = feature_extractor(test_set2, roi)
test_set3 = feature_extractor(test_set3, roi)
test_set4 = feature_extractor(test_set4, roi)

In [224]:
test_set_1 = pd.DataFrame(test_set)
test_set_2 = pd.DataFrame(test_set2)
test_set_3 = pd.DataFrame(test_set3)
test_set_4 = pd.DataFrame(test_set4)

In [225]:
test_total_day = pd.concat([test_set_1, test_set_2], axis=1)
test_total_wet = pd.concat([test_set_3, test_set_4], axis=1)

In [230]:
test_total_day

0           1         2           3         4           5   \
0    0.001177  126.487147  0.528630  134.194262  0.148640  235.159476   
1    0.002038   67.819527  0.587500   82.228733  0.217307  220.720494   
2    0.002195   52.693917  0.675396   81.178423  0.234238  230.797036   
3    0.001507   54.209811  0.789690  128.907160  0.233338  224.403485   
4    0.003075   46.993280  0.643351   65.928291  0.256584  259.561701   
..        ...         ...       ...         ...       ...         ...   
595  0.000818   44.986600  0.933528  338.355625  0.200443  206.967452   
596  0.000600   79.723816  0.890950  365.485839  0.154084  237.321841   
597  0.000770   60.436411  0.925661  406.508181  0.186653  213.123974   
598  0.000964   40.820124  0.928285  284.705610  0.209291  223.248613   
599  0.001106   32.937538  0.951518  339.639822  0.240942  199.126214   

              6         7          8         9   ...        14        15  \
0     410.289899  6.238696  10.437258  0.000364  ...  0.083156  0.043067   
1     261.095406  6.012363   9.776505  0.000531  ...  0.091289  0.045689   
2     272.019776  6.051516   9.638600  0.000602  ...  0.097156  0.048667   
3     461.418828  6.452288  10.068788  0.000595  ...  0.086622  0.040044   
4     216.719883  5.822552   9.288258  0.000601  ...  0.087911  0.037911   
..           ...       ...        ...       ...  ...       ...       ...   
595  1308.435899  7.049965  10.690372  0.000545  ...  0.078933  0.045333   
596  1382.219539  7.078614  11.132810  0.000359  ...  0.066622  0.036667   
597  1565.596313  7.085678  10.881493  0.000442  ...  0.064400  0.036356   
598  1098.002317  6.947337  10.528167  0.000544  ...  0.063822  0.034889   
599  1325.621751  6.942144  10.354730  0.000647  ...  0.056400  0.033156   

           16        17        18        19        20        21        22  \
0    0.041111  0.042000  0.060933  0.061067  0.124311  0.121689  0.348889   
1    0.043378  0.048222  0.062000  0.062044  0.124933  0.104356  0.334933   
2    0.038267  0.053511  0.056311  0.053467  0.111644  0.100578  0.351244   
3    0.041733  0.046978  0.068578  0.064444  0.118489  0.107333  0.345333   
4    0.035289  0.039422  0.057956  0.061511  0.130489  0.112667  0.353244   
..        ...       ...       ...       ...       ...       ...       ...   
595  0.051644  0.055956  0.084622  0.070933  0.121911  0.122133  0.302667   
596  0.051378  0.060267  0.102444  0.077867  0.134889  0.141600  0.270756   
597  0.049956  0.055289  0.106000  0.073511  0.132978  0.173644  0.254400   
598  0.050267  0.050533  0.097556  0.082844  0.139778  0.171289  0.256533   
599  0.050222  0.055022  0.103422  0.081467  0.147422  0.177378  0.245600   

        23  
0    130.0  
1    117.0  
2    130.0  
3    131.0  
4    139.0  
..     ...  
595  127.0  
596  139.0  
597  126.0  
598  126.0  
599  118.0  

[600 rows x 48 columns]

In [148]:
pred = svc_grid.predict(test_set2)

In [149]:
tester(pred, 3)

accuracy: 76.00


In [69]:
dd_total = pd.concat([dd1_df, dd3_df], axis = 1)

In [75]:
dd_total.drop('target', axis=1, inplace=True)

In [76]:
wd_total = pd.concat([wd1_df, wd3_df], axis=1)
wd_total.drop('target', axis=1, inplace=True)

In [77]:
dd_total['target'] = 0
wd_total['target'] = 1

In [33]:
pd.concat([dd1_df, dd3_df, wd1_df, wd3_df], axis = 0)

h_1        h_2       h_3        h_4       h_5         h_6  \
0     0.004249  39.600540  0.616880  51.724330  0.277161  278.715251   
1     0.002704  43.131528  0.772732  94.900740  0.267291  262.505695   
2     0.001660  50.052290  0.717742  88.706920  0.198975  252.671326   
3     0.002780  43.777802  0.585919  52.864535  0.227660  234.432181   
4     0.002051  42.213222  0.701762  70.789492  0.228693  260.102790   
...        ...        ...       ...        ...       ...         ...   
2035  0.048550   0.724741  0.965080  10.382225  0.810073  225.950363   
2036  0.064111   0.509752  0.959254   6.257159  0.844831  238.825757   
2037  0.053889   0.518441  0.965726   7.564670  0.838675  225.456232   
2038  0.041117   1.001205  0.959527  12.374793  0.800030  237.029934   
2039  0.022749   1.094867  0.982238  30.823369  0.762990  223.910250   

             h_7       h_8       h_9      h_10  ...     lbp_2     lbp_3  \
0     167.296781  5.567350  8.895602  0.000626  ...  0.107467  0.028667   
1     336.471433  6.084120  9.507177  0.000632  ...  0.107467  0.028889   
2     304.775388  6.114454  9.811530  0.000514  ...  0.119956  0.016444   
3     167.680337  5.702623  9.265208  0.000598  ...  0.117556  0.021556   
4     240.944748  5.980316  9.520294  0.000572  ...  0.117022  0.024978   
...          ...       ...       ...       ...  ...       ...       ...   
2035   40.804159  4.473791  5.138576  0.003850  ...  0.032933  0.032444   
2036   24.518885  4.144152  4.629738  0.004239  ...  0.027111  0.023689   
2037   29.740240  4.285750  4.798961  0.004492  ...  0.032222  0.025067   
2038   48.497968  4.682723  5.415121  0.003662  ...  0.034178  0.029467   
2039  122.198611  5.301251  6.182306  0.003400  ...  0.034444  0.037289   

         lbp_4     lbp_5     lbp_6     lbp_7     lbp_8     lbp_9    lbp_10  \
0     0.016533  0.024400  0.031156  0.037689  0.116400  0.118044  0.391556   
1     0.018978  0.021733  0.026711  0.032311  0.110978  0.129200  0.393200   
2     0.006800  0.011600  0.020622  0.017244  0.098444  0.149067  0.411822   
3     0.009689  0.015378  0.021556  0.021289  0.104489  0.142622  0.404800   
4     0.011556  0.019778  0.025289  0.024667  0.106667  0.132978  0.402400   
...        ...       ...       ...       ...       ...       ...       ...   
2035  0.078222  0.091378  0.173956  0.100267  0.125867  0.200400  0.145067   
2036  0.081867  0.092756  0.182578  0.101556  0.125778  0.230178  0.120889   
2037  0.073778  0.086133  0.190933  0.111556  0.119244  0.221733  0.124444   
2038  0.080978  0.087822  0.171511  0.102356  0.123156  0.210756  0.139822   
2039  0.089156  0.101067  0.176489  0.099467  0.112400  0.161422  0.164400   

      bright  
0      149.0  
1      140.0  
2      135.0  
3      126.0  
4      140.0  
...      ...  
2035   121.0  
2036   128.0  
2037   121.0  
2038   127.0  
2039   120.0  

[8160 rows x 24 columns]

In [42]:
def pca_ratio(df, count):
    for i in range(1, count+1):
        pca = PCA(n_components=i, random_state=2021)
        fit_pca = pca.fit(df)
        print(f'{i}: {sum(fit_pca.explained_variance_ratio_)*100:.4f}%')

In [47]:
pca_ratio(wd1_df, 20)

1: 79.2869%
2: 96.6034%
3: 99.9835%
4: 99.9934%
5: 99.9988%
6: 99.9999%
7: 99.9999%
8: 100.0000%
9: 100.0000%
10: 100.0000%
11: 100.0000%
12: 100.0000%
13: 100.0000%
14: 100.0000%
15: 100.0000%
16: 100.0000%
17: 100.0000%
18: 100.0000%
19: 100.0000%
20: 100.0000%


In [48]:
pca_ratio(wd3_df, 20)

1: 73.4370%
2: 99.8102%
3: 99.9833%
4: 99.9922%
5: 99.9981%
6: 99.9999%
7: 99.9999%
8: 100.0000%
9: 100.0000%
10: 100.0000%
11: 100.0000%
12: 100.0000%
13: 100.0000%
14: 100.0000%
15: 100.0000%
16: 100.0000%
17: 100.0000%
18: 100.0000%
19: 100.0000%
20: 100.0000%


In [49]:
pca_ratio(wd1_df_ol, 20)

1: 74.0935%
2: 95.6287%
3: 99.5760%
4: 99.9090%
5: 99.9780%
6: 99.9981%
7: 99.9994%
8: 99.9997%
9: 99.9998%
10: 99.9999%
11: 100.0000%
12: 100.0000%
13: 100.0000%
14: 100.0000%
15: 100.0000%
16: 100.0000%
17: 100.0000%
18: 100.0000%
19: 100.0000%
20: 100.0000%


In [50]:
pca_ratio(wd3_df_ol, 20)

1: 68.3208%
2: 99.1447%
3: 99.5330%
4: 99.8259%
5: 99.9613%
6: 99.9877%
7: 99.9958%
8: 99.9991%
9: 99.9995%
10: 99.9997%
11: 99.9998%
12: 99.9999%
13: 99.9999%
14: 99.9999%
15: 100.0000%
16: 100.0000%
17: 100.0000%
18: 100.0000%
19: 100.0000%
20: 100.0000%


In [20]:
pca = PCA(n_components=10, random_state=2021)

In [21]:
pca_fit = pca.fit_transform(wd1_df)
wd1_pca_df = pd.DataFrame(pca_fit)
pca_fit = pca.fit_transform(wd3_df)
wd3_pca_df = pd.DataFrame(pca_fit)

In [22]:
t_test2(wd1_pca_df, wd3_pca_df)

0 = t_value: 0.0000, p_value: 1.0000
1 = t_value: 0.0000, p_value: 1.0000
2 = t_value: 0.0000, p_value: 1.0000
3 = t_value: -0.0000, p_value: 1.0000
4 = t_value: 0.0000, p_value: 1.0000
5 = t_value: 0.0000, p_value: 1.0000
6 = t_value: -0.0000, p_value: 1.0000
7 = t_value: 0.0000, p_value: 1.0000
8 = t_value: -0.0000, p_value: 1.0000
9 = t_value: 0.0000, p_value: 1.0000
0.05이하 채택 된 dict_keys([])


In [23]:
wd1_no_pca_df = no.normalizer(wd1_pca_df)
wd3_no_pca_df = no.normalizer(wd3_pca_df)

In [24]:
t_test2(wd1_no_pca_df, wd3_no_pca_df)

0 = t_value: -0.0000, p_value: 1.0000
1 = t_value: 0.0000, p_value: 1.0000
2 = t_value: 0.0000, p_value: 1.0000
3 = t_value: 0.0000, p_value: 1.0000
4 = t_value: -0.0000, p_value: 1.0000
5 = t_value: -0.0000, p_value: 1.0000
6 = t_value: 0.0000, p_value: 1.0000
7 = t_value: -0.0000, p_value: 1.0000
8 = t_value: 0.0000, p_value: 1.0000
9 = t_value: 0.0000, p_value: 1.0000
0.05이하 채택 된 dict_keys([])


In [25]:
pca_fit = pca.fit_transform(wd1_df_ol)
wd1_ol_pca_df = pd.DataFrame(pca_fit)
pca_fit = pca.fit_transform(wd3_df_ol)
wd3_ol_pca_df = pd.DataFrame(pca_fit)

In [26]:
wd1_olno_pca_df = no.normalizer(wd1_ol_pca_df)
wd3_olno_pca_df = no.normalizer(wd3_ol_pca_df)

In [27]:
t_test2(wd1_olno_pca_df, wd3_olno_pca_df)

0 = t_value: 0.0000, p_value: 1.0000
1 = t_value: 0.0000, p_value: 1.0000
2 = t_value: 0.0000, p_value: 1.0000
3 = t_value: 0.0000, p_value: 1.0000
4 = t_value: 0.0000, p_value: 1.0000
5 = t_value: 0.0000, p_value: 1.0000
6 = t_value: 0.0000, p_value: 1.0000
7 = t_value: -0.0000, p_value: 1.0000
8 = t_value: -0.0000, p_value: 1.0000
9 = t_value: -0.0000, p_value: 1.0000
0.05이하 채택 된 dict_keys([])


In [28]:
anova_test(wd1_ol_pca_df, wd3_ol_pca_df)

0 = f_value: 0.0000, p_value: 1.0000
1 = f_value: 0.0000, p_value: 1.0000
2 = f_value: 0.0000, p_value: 1.0000
3 = f_value: 0.0000, p_value: 1.0000
4 = f_value: 0.0000, p_value: 1.0000
5 = f_value: 0.0000, p_value: 1.0000
6 = f_value: 0.0000, p_value: 1.0000
7 = f_value: 0.0000, p_value: 1.0000
8 = f_value: 0.0000, p_value: 1.0000
9 = f_value: 0.0000, p_value: 1.0000


In [29]:
anova_test(wd1_olno_pca_df, wd3_olno_pca_df)

0 = f_value: 0.0000, p_value: 1.0000
1 = f_value: 0.0000, p_value: 1.0000
2 = f_value: 0.0000, p_value: 1.0000
3 = f_value: 0.0000, p_value: 1.0000
4 = f_value: 0.0000, p_value: 1.0000
5 = f_value: 0.0000, p_value: 1.0000
6 = f_value: 0.0000, p_value: 1.0000
7 = f_value: 0.0000, p_value: 1.0000
8 = f_value: 0.0000, p_value: 1.0000
9 = f_value: 0.0000, p_value: 1.0000


In [30]:
anova_test(wd1_pca_df, wd3_pca_df)

0 = f_value: 0.0000, p_value: 1.0000
1 = f_value: 0.0000, p_value: 1.0000
2 = f_value: 0.0000, p_value: 1.0000
3 = f_value: 0.0000, p_value: 1.0000
4 = f_value: 0.0000, p_value: 1.0000
5 = f_value: 0.0000, p_value: 1.0000
6 = f_value: 0.0000, p_value: 1.0000
7 = f_value: 0.0000, p_value: 1.0000
8 = f_value: 0.0000, p_value: 1.0000
9 = f_value: 0.0000, p_value: 1.0000


In [31]:
anova_test(wd1_ol_pca_df, wd3_ol_pca_df)

0 = f_value: 0.0000, p_value: 1.0000
1 = f_value: 0.0000, p_value: 1.0000
2 = f_value: 0.0000, p_value: 1.0000
3 = f_value: 0.0000, p_value: 1.0000
4 = f_value: 0.0000, p_value: 1.0000
5 = f_value: 0.0000, p_value: 1.0000
6 = f_value: 0.0000, p_value: 1.0000
7 = f_value: 0.0000, p_value: 1.0000
8 = f_value: 0.0000, p_value: 1.0000
9 = f_value: 0.0000, p_value: 1.0000


In [35]:
wd1_df['target'] = 0

In [36]:
wd3_df['target'] = 1

In [32]:
pd.concat([wd1_df, wd3_df], axis=1)

h_1        h_2       h_3        h_4       h_5         h_6  \
0     0.006427  18.650207  0.620961  24.610744  0.335934  181.146657   
1     0.004892  19.684701  0.696239  32.412066  0.325626  175.877509   
2     0.005899  20.652577  0.598675  25.730500  0.322048  193.238126   
3     0.005364  19.342807  0.651989  27.790615  0.321498  186.213668   
4     0.004778  19.523510  0.740892  37.684322  0.350071  194.281631   
...        ...        ...       ...        ...       ...         ...   
2041  0.017448   6.756051  0.800598  16.940586  0.576759  228.280113   
2042  0.031746   3.107798  0.815274   8.418836  0.634757  241.500692   
2043  0.029947   2.575771  0.847585   8.449651  0.650453  233.940591   
2044  0.027350   1.375624  0.947626  13.132259  0.722587  242.781399   
2045  0.033725   1.470899  0.923100   9.569890  0.722990  229.125034   

             h_7       h_8       h_9      h_10  ...     lbp_2     lbp_3  \
0      79.792769  5.182362  8.121547  0.001177  ...  0.104756  0.040978   
1     109.963562  5.424743  8.428748  0.001134  ...  0.106267  0.041289   
2      82.269422  5.200111  8.208437  0.000989  ...  0.106489  0.041556   
3      91.819654  5.292080  8.283429  0.001052  ...  0.107689  0.042044   
4     131.213779  5.546547  8.469450  0.001134  ...  0.101511  0.044533   
...          ...       ...       ...       ...  ...       ...       ...   
2041   61.006293  4.961762  6.803908  0.001888  ...  0.032933  0.032444   
2042   30.567548  4.442000  5.876937  0.002370  ...  0.027111  0.023689   
2043   31.222831  4.499991  5.881482  0.002601  ...  0.032222  0.025067   
2044   51.153413  4.775089  5.831177  0.002982  ...  0.034178  0.029467   
2045   36.808663  4.594691  5.624009  0.003142  ...  0.034444  0.037289   

         lbp_4     lbp_5     lbp_6     lbp_7     lbp_8     lbp_9    lbp_10  \
0     0.033244  0.033022  0.047333  0.047689  0.113689  0.115378  0.363511   
1     0.030000  0.029289  0.041956  0.046133  0.116089  0.115689  0.369156   
2     0.029778  0.030222  0.044667  0.041644  0.110711  0.120978  0.372178   
3     0.032400  0.030667  0.044622  0.042178  0.115200  0.117422  0.364489   
4     0.038622  0.032711  0.047778  0.051333  0.112711  0.109778  0.362578   
...        ...       ...       ...       ...       ...       ...       ...   
2041  0.078222  0.091378  0.173956  0.100267  0.125867  0.200400  0.145067   
2042  0.081867  0.092756  0.182578  0.101556  0.125778  0.230178  0.120889   
2043  0.073778  0.086133  0.190933  0.111556  0.119244  0.221733  0.124444   
2044  0.080978  0.087822  0.171511  0.102356  0.123156  0.210756  0.139822   
2045  0.089156  0.101067  0.176489  0.099467  0.112400  0.161422  0.164400   

      bright  
0      104.0  
1      100.0  
2      109.0  
3      102.0  
4      102.0  
...      ...  
2041   121.0  
2042   128.0  
2043   121.0  
2044   127.0  
2045   120.0  

[2046 rows x 48 columns]

In [37]:
wd_concat = pd.concat([wd1_df, wd3_df], axis=0)
wd_concat.reset_index(inplace=True, drop=True)

In [34]:
wd_concat['target'] = 0

In [75]:
x_train = wd_concat.iloc[:, :-1]
y_train = wd_concat.iloc[:, -1]

In [38]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [45]:
lda = LinearDiscriminantAnalysis(n_components=1, solver="svd", store_covariance=True).fit(x_train, y_train)

In [58]:
test_lda = pd.DataFrame(lda.transform(x_train), columns = ['lda_component_1'])

In [59]:
test_lda['target'] = y_train

In [62]:
markers = ['^', 's']

In [64]:
x_axis_data = test_lda[test_lda['target']==0]['lda_component_1']
y_axis_data = test_lda[test_lda['target'] ==1]['lda_component_1']

In [76]:
x_train, x_test, y_train, y_test = train_test_split(x_train, y_train, test_size=0.2, random_state=2021)

In [77]:
lda = LinearDiscriminantAnalysis(n_components=1, solver="svd", store_covariance=True).fit(x_train, y_train)

In [78]:
lda.fit(x_train, y_train)

LinearDiscriminantAnalysis(n_components=1, store_covariance=True)

In [79]:
pred = lda.predict(x_test)

In [84]:
from sklearn.metrics import classification_report, accuracy_score

In [83]:
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.97      0.95      0.96       411
           1       0.95      0.97      0.96       408

    accuracy                           0.96       819
   macro avg       0.96      0.96      0.96       819
weighted avg       0.96      0.96      0.96       819



In [85]:
accuracy_score(y_test, pred)

0.960927960927961

In [46]:
lda.means_

array([[ 1.26950884e-02,  1.30263755e+01,  7.19796677e-01,
         2.14113156e+01,  4.41057437e-01,  2.36046038e+02,
         7.26188870e+01,  5.01903421e+00,  7.42913712e+00,
         1.49905609e-03,  2.69871636e+00, -2.09120899e-01,
         8.45507429e-01,  9.74583806e-02,  9.81800271e-02,
         3.45773181e-02,  3.09227948e-02,  3.64544984e-02,
         5.37597644e-02,  4.81312866e-02,  1.21397099e-01,
         1.32520678e-01,  3.46598054e-01,  1.26608993e+02],
       [ 4.89610736e-02,  1.36628543e+00,  9.29674110e-01,
         1.10753689e+01,  7.87599091e-01,  2.23222432e+02,
         4.29351902e+01,  4.42264145e+00,  5.22338770e+00,
         3.98329387e-03,  1.32832022e+00, -5.44522312e-01,
         9.82912118e-01,  2.71154529e-02,  4.02901448e-02,
         2.82385548e-02,  6.54747192e-02,  7.19081279e-02,
         1.52315897e-01,  9.23697312e-02,  1.31415366e-01,
         2.17725383e-01,  1.73146523e-01,  1.19968231e+02]])

In [47]:
lda.covariance_

array([[ 2.50066869e-04, -4.36721403e-02,  5.33789574e-04,
        -8.59518683e-02,  1.20447391e-03, -5.30035688e-02,
        -3.00135333e-01, -4.85337580e-03, -1.06550602e-02,
         1.14771702e-05, -5.18551009e-03, -9.94324038e-04,
         3.82679236e-04, -2.28006556e-04, -1.77906076e-04,
        -2.52003094e-05,  8.17680770e-05,  7.77086871e-05,
         3.08707201e-04,  1.23139274e-04, -3.83740547e-06,
         4.80256671e-04, -6.36629563e-04, -2.81065491e-02],
       [-4.36721403e-02,  5.63298524e+01, -6.14625672e-01,
         4.49498850e+01, -7.14803699e-01,  2.11433033e+00,
         1.23469688e+02,  1.31718613e+00,  4.86968689e+00,
        -3.43479623e-03,  3.06652391e+00,  5.32837320e-01,
        -4.87904794e-01,  1.41218967e-01,  7.58020096e-02,
        -3.04468300e-02, -7.62755140e-02, -7.93421984e-02,
        -1.36466920e-01, -9.64491057e-02, -2.60632307e-02,
        -2.02318788e-02,  2.48254701e-01,  1.50555963e+00],
       [ 5.33789574e-04, -6.14625672e-01,  1.11461672e

In [48]:
import matplotlib.pyplot as plt

In [93]:
svc = SVC(C=10, kernel='linear')

In [94]:
svc.fit(x_train, y_train)

SVC(C=10, kernel='linear')

In [106]:
test_set = feature_extractor('D:/road_201_203', roi)

In [107]:
pred = svc.predict(test_set)

In [108]:
tester(pred, 0)

accuracy: 97.75


In [109]:
test_set2 = feature_extractor('D:/road_203', roi)

In [110]:
pred = svc.predict(test_set2)

In [111]:
tester(pred, 1)

accuracy: 90.54


In [100]:
svc.predict(test_set[6:])

array([1, 1, 0, 1, 1, 1], dtype=int64)

In [18]:
wd_test_df = pd.DataFrame(wd_test, columns=columns)

In [20]:
wd_test_df

h_1        h_2       h_3        h_4       h_5         h_6        h_7  \
0   0.020892  10.865464  0.354792   8.420061  0.461682  150.443145  22.814781   
1   0.017627  10.796972  0.442616   9.685393  0.447906  147.188929  27.944600   
2   0.012769  13.952655  0.410883  11.841282  0.376514  156.195649  33.412475   
3   0.016393  12.601763  0.378142  10.131482  0.406476  151.995859  27.924167   
4   0.018460   8.475243  0.668025  12.764846  0.528469  157.960580  42.584141   
5   0.039276   6.200956  0.530837   6.607195  0.586933  147.591288  20.227824   
6   0.166116   0.272969  0.912454   1.559663  0.909265  141.362073   5.965685   
7   0.110263   1.486196  0.704585   2.515722  0.830118  137.423910   8.576692   
8   0.150885   0.389729  0.875853   1.570528  0.880662  145.946438   5.892381   
9   0.113658   1.288128  0.808057   3.356358  0.878066  139.361008  12.137305   
10  0.111058   0.374245  0.955778   4.232360  0.904968  141.223920  16.555195   
11  0.198779   0.222500  0.895662   1.066728  0.917371  134.262090   4.044411   

         h_8       h_9      h_10  ...     lbp_2     lbp_3     lbp_4     lbp_5  \
0   4.213453  6.603844  0.001999  ...  0.096311  0.029511  0.021067  0.021244   
1   4.389639  6.842849  0.002051  ...  0.101822  0.028667  0.018800  0.018800   
2   4.497402  7.185831  0.001654  ...  0.104711  0.031867  0.017200  0.019556   
3   4.350695  6.939075  0.001814  ...  0.103467  0.029600  0.019556  0.021556   
4   4.700233  6.834916  0.002236  ...  0.083378  0.032222  0.027378  0.030178   
5   4.097195  5.945998  0.003042  ...  0.077289  0.026978  0.028489  0.029467   
6   2.920356  3.220585  0.008991  ...  0.022978  0.009556  0.041911  0.025467   
7   3.331794  4.068997  0.006606  ...  0.034578  0.013244  0.050000  0.028933   
8   3.015989  3.421178  0.008038  ...  0.030000  0.016489  0.050000  0.040578   
9   3.418661  3.911821  0.007449  ...  0.022267  0.013022  0.054044  0.044133   
10  3.538411  3.850628  0.008886  ...  0.023467  0.009422  0.040178  0.032133   
11  2.620791  2.871173  0.009859  ...  0.016356  0.007644  0.041289  0.031111   

       lbp_6     lbp_7     lbp_8     lbp_9    lbp_10  bright  
0   0.040178  0.038711  0.126133  0.142978  0.385333    81.0  
1   0.037644  0.038533  0.128400  0.141600  0.381244    80.0  
2   0.032667  0.029867  0.118622  0.130267  0.404756    85.0  
3   0.033422  0.036622  0.122800  0.131156  0.395511    83.0  
4   0.056178  0.051289  0.138667  0.146844  0.350622    86.0  
5   0.060000  0.056622  0.150400  0.153467  0.340356    80.0  
6   0.133022  0.068311  0.128000  0.467733  0.089822    77.0  
7   0.133733  0.067911  0.134800  0.338400  0.172489    75.0  
8   0.132800  0.073733  0.141733  0.365867  0.134133    79.0  
9   0.145644  0.073244  0.123600  0.401867  0.107156    75.0  
10  0.138889  0.075867  0.136356  0.440311  0.091867    76.0  
11  0.148933  0.064756  0.127911  0.465200  0.086133    72.0  

[12 rows x 24 columns]

In [21]:
wd_test_df_201 = wd_test_df[:6]
wd_test_df_203 = wd_test_df[6:]

In [22]:
wd_test_df_201

h_1        h_2       h_3        h_4       h_5         h_6        h_7  \
0  0.020892  10.865464  0.354792   8.420061  0.461682  150.443145  22.814781   
1  0.017627  10.796972  0.442616   9.685393  0.447906  147.188929  27.944600   
2  0.012769  13.952655  0.410883  11.841282  0.376514  156.195649  33.412475   
3  0.016393  12.601763  0.378142  10.131482  0.406476  151.995859  27.924167   
4  0.018460   8.475243  0.668025  12.764846  0.528469  157.960580  42.584141   
5  0.039276   6.200956  0.530837   6.607195  0.586933  147.591288  20.227824   

        h_8       h_9      h_10  ...     lbp_2     lbp_3     lbp_4     lbp_5  \
0  4.213453  6.603844  0.001999  ...  0.096311  0.029511  0.021067  0.021244   
1  4.389639  6.842849  0.002051  ...  0.101822  0.028667  0.018800  0.018800   
2  4.497402  7.185831  0.001654  ...  0.104711  0.031867  0.017200  0.019556   
3  4.350695  6.939075  0.001814  ...  0.103467  0.029600  0.019556  0.021556   
4  4.700233  6.834916  0.002236  ...  0.083378  0.032222  0.027378  0.030178   
5  4.097195  5.945998  0.003042  ...  0.077289  0.026978  0.028489  0.029467   

      lbp_6     lbp_7     lbp_8     lbp_9    lbp_10  bright  
0  0.040178  0.038711  0.126133  0.142978  0.385333    81.0  
1  0.037644  0.038533  0.128400  0.141600  0.381244    80.0  
2  0.032667  0.029867  0.118622  0.130267  0.404756    85.0  
3  0.033422  0.036622  0.122800  0.131156  0.395511    83.0  
4  0.056178  0.051289  0.138667  0.146844  0.350622    86.0  
5  0.060000  0.056622  0.150400  0.153467  0.340356    80.0  

[6 rows x 24 columns]

In [23]:
wd_test_df_203

h_1       h_2       h_3       h_4       h_5         h_6        h_7  \
6   0.166116  0.272969  0.912454  1.559663  0.909265  141.362073   5.965685   
7   0.110263  1.486196  0.704585  2.515722  0.830118  137.423910   8.576692   
8   0.150885  0.389729  0.875853  1.570528  0.880662  145.946438   5.892381   
9   0.113658  1.288128  0.808057  3.356358  0.878066  139.361008  12.137305   
10  0.111058  0.374245  0.955778  4.232360  0.904968  141.223920  16.555195   
11  0.198779  0.222500  0.895662  1.066728  0.917371  134.262090   4.044411   

         h_8       h_9      h_10  ...     lbp_2     lbp_3     lbp_4     lbp_5  \
6   2.920356  3.220585  0.008991  ...  0.022978  0.009556  0.041911  0.025467   
7   3.331794  4.068997  0.006606  ...  0.034578  0.013244  0.050000  0.028933   
8   3.015989  3.421178  0.008038  ...  0.030000  0.016489  0.050000  0.040578   
9   3.418661  3.911821  0.007449  ...  0.022267  0.013022  0.054044  0.044133   
10  3.538411  3.850628  0.008886  ...  0.023467  0.009422  0.040178  0.032133   
11  2.620791  2.871173  0.009859  ...  0.016356  0.007644  0.041289  0.031111   

       lbp_6     lbp_7     lbp_8     lbp_9    lbp_10  bright  
6   0.133022  0.068311  0.128000  0.467733  0.089822    77.0  
7   0.133733  0.067911  0.134800  0.338400  0.172489    75.0  
8   0.132800  0.073733  0.141733  0.365867  0.134133    79.0  
9   0.145644  0.073244  0.123600  0.401867  0.107156    75.0  
10  0.138889  0.075867  0.136356  0.440311  0.091867    76.0  
11  0.148933  0.064756  0.127911  0.465200  0.086133    72.0  

[6 rows x 24 columns]

In [24]:
wd_test_df_201.describe().T

count        mean       std         min         25%         50%  \
h_1       6.0    0.020903  0.009389    0.012769    0.016702    0.018044   
h_2       6.0   10.482175  2.797838    6.200956    9.055675   10.831218   
h_3       6.0    0.464216  0.117192    0.354792    0.386328    0.426750   
h_4       6.0    9.908377  2.240889    6.607195    8.736394    9.908438   
h_5       6.0    0.467997  0.077975    0.376514    0.416834    0.454794   
h_6       6.0  151.895908  4.428858  147.188929  148.304252  151.219502   
h_7       6.0   29.151331  8.016741   20.227824   24.092128   27.934383   
h_8       6.0    4.374770  0.212055    4.097195    4.247763    4.370167   
h_9       6.0    6.725419  0.425602    5.945998    6.661612    6.838883   
h_10      6.0    0.002133  0.000488    0.001654    0.001860    0.002025   
h_11      6.0    2.713278  0.240783    2.316787    2.606452    2.767549   
h_12      6.0   -0.119695  0.065380   -0.217358   -0.163955   -0.092137   
h_13      6.0    0.718264  0.123431    0.598595    0.622019    0.683417   
lbp_1     6.0    0.096667  0.013552    0.076933    0.087067    0.101511   
lbp_2     6.0    0.094496  0.011502    0.077289    0.086611    0.099067   
lbp_3     6.0    0.029807  0.001975    0.026978    0.028878    0.029556   
lbp_4     6.0    0.022081  0.004714    0.017200    0.018989    0.020311   
lbp_5     6.0    0.023467  0.005034    0.018800    0.019978    0.021400   
lbp_6     6.0    0.043348  0.011807    0.032667    0.034478    0.038911   
lbp_7     6.0    0.041941  0.009990    0.029867    0.037100    0.038622   
lbp_8     6.0    0.130837  0.011712    0.118622    0.123633    0.127267   
lbp_9     6.0    0.141052  0.009011    0.130267    0.133767    0.142289   
lbp_10    6.0    0.376304  0.025441    0.340356    0.358278    0.383289   
bright    6.0   82.500000  2.588436   80.000000   80.250000   82.000000   

               75%         max  
h_1       0.020284    0.039276  
h_2      12.167688   13.952655  
h_3       0.508782    0.668025  
h_4      11.413832   12.764846  
h_5       0.511773    0.586933  
h_6     155.145701  157.960580  
h_7      32.045506   42.584141  
h_8       4.470461    4.700233  
h_9       6.915018    7.185831  
h_10      0.002190    0.003042  
h_11      2.873409    2.966330  
h_12     -0.071075   -0.064590  
h_13      0.804408    0.897861  
lbp_1     0.105856    0.110489  
lbp_2     0.103056    0.104711  
lbp_3     0.031300    0.032222  
lbp_4     0.025800    0.028489  
lbp_5     0.027489    0.030178  
lbp_6     0.052178    0.060000  
lbp_7     0.048144    0.056622  
lbp_8     0.136100    0.150400  
lbp_9     0.145878    0.153467  
lbp_10    0.392967    0.404756  
bright   84.500000   86.000000

In [25]:
wd_test_df_203.describe().T

count        mean       std         min         25%         50%  \
h_1       6.0    0.141793  0.036475    0.110263    0.111708    0.132271   
h_2       6.0    0.672294  0.560738    0.222500    0.298288    0.381987   
h_3       6.0    0.858732  0.089767    0.704585    0.825006    0.885758   
h_4       6.0    2.383560  1.221126    1.066728    1.562379    2.043125   
h_5       6.0    0.886742  0.031925    0.830118    0.878715    0.892815   
h_6       6.0  139.929906  3.966158  134.262090  137.908184  140.292464   
h_7       6.0    8.861945  4.699164    4.044411    5.910707    7.271188   
h_8       6.0    3.141000  0.348242    2.620791    2.944264    3.173891   
h_9       6.0    3.557397  0.463948    2.871173    3.270733    3.635903   
h_10      6.0    0.008305  0.001176    0.006606    0.007596    0.008462   
h_11      6.0    0.905896  0.208971    0.700930    0.766073    0.852049   
h_12      6.0   -0.588012  0.078599   -0.690666   -0.621263   -0.604679   
h_13      6.0    0.969221  0.014941    0.953429    0.958442    0.965462   
lbp_1     6.0    0.015163  0.005535    0.010667    0.011933    0.013933   
lbp_2     6.0    0.024941  0.006409    0.016356    0.022444    0.023222   
lbp_3     6.0    0.011563  0.003262    0.007644    0.009456    0.011289   
lbp_4     6.0    0.046237  0.005817    0.040178    0.041444    0.045956   
lbp_5     6.0    0.033726  0.007153    0.025467    0.029478    0.031622   
lbp_6     6.0    0.138837  0.006994    0.132800    0.133200    0.136311   
lbp_7     6.0    0.070637  0.004270    0.064756    0.068011    0.070778   
lbp_8     6.0    0.132067  0.006705    0.123600    0.127933    0.131400   
lbp_9     6.0    0.413230  0.053622    0.338400    0.374867    0.421089   
lbp_10    6.0    0.113600  0.033832    0.086133    0.090333    0.099511   
bright    6.0   75.666667  2.338090   72.000000   75.000000   75.500000   

               75%         max  
h_1       0.162308    0.198779  
h_2       1.063528    1.486196  
h_3       0.908256    0.955778  
h_4       3.146199    4.232360  
h_5       0.908191    0.917371  
h_6     141.327535  145.946438  
h_7      11.247151   16.555195  
h_8       3.396944    3.538411  
h_9       3.896523    4.068997  
h_10      0.008965    0.009859  
h_11      0.980865    1.269127  
h_12     -0.557126   -0.459985  
h_13      0.979174    0.991127  
lbp_1     0.014933    0.025911  
lbp_2     0.028367    0.034578  
lbp_3     0.013189    0.016489  
lbp_4     0.050000    0.054044  
lbp_5     0.038467    0.044133  
lbp_6     0.143956    0.148933  
lbp_7     0.073611    0.075867  
lbp_8     0.135967    0.141733  
lbp_9     0.458978    0.467733  
lbp_10    0.127389    0.172489  
bright   76.750000   79.000000

In [26]:
t_test2(wd_test_df_201, wd_test_df_203)

h_1 = t_value: -7.8621, p_value: 0.0000
h_2 = t_value: 8.4210, p_value: 0.0000
h_3 = t_value: -6.5462, p_value: 0.0001
h_4 = t_value: 7.2225, p_value: 0.0000
h_5 = t_value: -12.1736, p_value: 0.0000
h_6 = t_value: 4.9301, p_value: 0.0006
h_7 = t_value: 5.3483, p_value: 0.0003
h_8 = t_value: 7.4121, p_value: 0.0000
h_9 = t_value: 12.3255, p_value: 0.0000
h_10 = t_value: -11.8715, p_value: 0.0000
h_11 = t_value: 13.8861, p_value: 0.0000
h_12 = t_value: 11.2204, p_value: 0.0000
h_13 = t_value: -4.9441, p_value: 0.0006
lbp_1 = t_value: 13.6377, p_value: 0.0000
lbp_2 = t_value: 12.9397, p_value: 0.0000
lbp_3 = t_value: 11.7199, p_value: 0.0000
lbp_4 = t_value: -7.9025, p_value: 0.0000
lbp_5 = t_value: -2.8731, p_value: 0.0166
lbp_6 = t_value: -17.0437, p_value: 0.0000
lbp_7 = t_value: -6.4697, p_value: 0.0001
lbp_8 = t_value: -0.2232, p_value: 0.8279
lbp_9 = t_value: -12.2613, p_value: 0.0000
lbp_10 = t_value: 15.2016, p_value: 0.0000
bright = t_value: 4.7987, p_value: 0.0007
0.05이하 채택 된 di

In [28]:
anova_test(wd_test_df_201, wd_test_df_203)

차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
차이가 있음
lbp_8 = f_value: 0.0498, p_value: 0.8279
차이가 있음
차이가 있음
차이가 있음


In [29]:
source_path = 'D:/road_train/wet_day'
source_path_203 = 'D:/road_train/wet_day_203'

In [53]:
import fnmatch
import os
import shutil

In [39]:
img_201 = fnmatch.filter(os.listdir(source_path), '*.jpg')

In [33]:
img_203 = fnmatch.filter(os.listdir(source_path_203), '*.jpg')

In [37]:
img_203[0].split("_")[1][:13]

'2020111907025'

In [54]:
aa = []
for img in img_203:
#     filter = img.split("_")[1][:13]
    ilters = fnmatch.filter(os.listdir(source_path), f"OPTICAL201_{img.split('_')[1][:12]}*.jpg")
    path = f'{source_path}/{ilters[0]}'
    aa.append(path)
    shutil.copy(path, f'D:/road_train/wet_day_201/{ilters[0]}')

In [52]:
len(aa)

341

In [6]:
dry_day = feature_extractor('D:/road_train/dry_day', roi)
dry_night = feature_extractor('D:/road_train/dry_night', roi)
wet_day = feature_extractor('D:/road_train/wet_day', roi)
wet_night = feature_extractor('D:/road_train/wet_night', roi)

In [7]:
dd_df = pd.DataFrame(dry_day, columns=columns)
dn_df = pd.DataFrame(dry_night, columns=columns)
wd_df = pd.DataFrame(wet_day, columns=columns)
wn_df = pd.DataFrame(wet_night, columns=columns)

NameError: name 'columns' is not defined

In [ ]:
dd_df_ol = ol(dd_df)
dn_df_ol = ol(dn_df)
wd_df_ol = ol(wd_df)
wn_df_ol = ol(wn_df)

In [10]:
def ttest_2samp(df1,df2):
    
    p_value_05 = {}
    
    for x in df1.columns:
        t_value = stats.ttest_ind(df1[x], df2[x]).statistic
        p_value = stats.ttest_ind(df1[x], df2[x]).pvalue

        print(f'{x} = t_value: {t_value:.4f}, p_value: {p_value:.4f}')
        
        if p_value < 0.05:
            p_value_05[x] = p_value
            
    
    print(f'0.05이하 채택 된 {p_value_05.keys()}')

In [11]:
dd_df['target'] = 0
dn_df['target'] = 1
wd_df['target'] = 2
wn_df['target'] = 3

dd_df_ol['target'] = 0
dn_df_ol['target'] = 1
wd_df_ol['target'] = 2
wn_df_ol['target'] = 3

In [12]:
all_df = pd.concat([dd_df, dn_df, wd_df, wn_df], axis=0)
all_df.reset_index(inplace=True, drop=True)

In [13]:
dwd_df = pd.concat([dd_df, wd_df], axis=0)
dwd_df.reset_index(inplace=True, drop=True)

In [14]:
x_train, x_test, y_train, y_test = train_test_split(all_df.iloc[:, :-1], all_df.iloc[:, -1], test_size=0.2, random_state=2021)

In [15]:
svc = SVC(C = 15, kernel='poly', random_state=2021)

In [16]:
svc.fit(x_train, y_train)

SVC(C=15, kernel='poly', random_state=2021)

In [17]:
source_path = 'D:/Harry/Bright/MK-SD53R/dry_day'

In [18]:
source_path_1 = 'D:/Harry/Bright/MK-SD53R/wet_day'

In [19]:
source_path_2 = 'D:/Harry/Bright/MK-SD53R/dry_night'

In [20]:
source_path_3 = 'D:/Harry/Bright/MK-SD53R/wet_night'

In [21]:
test_sample = feature_extractor(source_path, roi)

In [22]:
test_sample_1 = feature_extractor(source_path_1, roi)

In [23]:
test_sample_2 = feature_extractor(source_path_2, roi)

In [24]:
test_sample_3 = feature_extractor(source_path_3, roi)

In [25]:
pred = svc.predict(test_sample)
pred_1 = svc.predict(test_sample_1)
pred_2 = svc.predict(test_sample_2)
pred_3 = svc.predict(test_sample_3)

In [26]:
tester(pred, 0)

accuracy: 64.44


In [27]:
tester(pred_1, 1)

accuracy: 0.00


In [28]:
tester(pred_2, 2)

accuracy: 0.00


In [29]:
tester(pred_3, 3)

accuracy: 33.61


dd = accuracy: 96.94, wd = accuracy: 24.17
dd = accuracy: 98.33; wd = accuracy: 28.61

dry_day = 82.50%; dry_night: 73.61%; wet_day = 0.83%; wet_night = 30.83%

In [30]:
kmeans = KMeans(n_clusters=4, max_iter=1000, random_state=2021)

In [31]:
kmeans.fit(all_df.iloc[:, :-1])

KMeans(max_iter=1000, n_clusters=4, random_state=2021)

In [32]:
kmeans.labels_

array([1, 1, 3, ..., 0, 1, 1])

In [33]:
all_df['kmeans'] = kmeans.labels_

In [34]:
all_df['target']

0        0
1        0
2        0
3        0
4        0
        ..
15995    3
15996    3
15997    3
15998    3
15999    3
Name: target, Length: 16000, dtype: int64

In [35]:
all_df['kmeans']

0        1
1        1
2        3
3        1
4        1
        ..
15995    0
15996    1
15997    0
15998    1
15999    1
Name: kmeans, Length: 16000, dtype: int32

In [36]:
all_df[:4000]['kmeans'].value_counts()

1    1934
3     880
0     775
2     411
Name: kmeans, dtype: int64

In [37]:
all_df[4000:]['kmeans'].value_counts()

0    8899
1    2611
3     489
2       1
Name: kmeans, dtype: int64

In [38]:
pca = PCA(n_components=2)

In [39]:
trans = pca.fit_transform(all_df.iloc[:, :-1])

In [40]:
trans[:, 0]

array([ 38.37952549,  87.11747954, 136.45497843, ..., -53.87170767,
        59.85892663,  -6.34366612])

In [41]:
all_df['pca_x'] = trans[:, 0]
all_df['pca_y'] = trans[:, 1]

In [42]:
mark0 = all_df[all_df['kmeans']==0].index
mark1 = all_df[all_df['kmeans']==1].index
mark2 = all_df[all_df['kmeans']==2].index
mark3 = all_df[all_df['kmeans']==3].index

In [43]:
# fig = go.Figure()
# fig.add_trace(go.Scatter(x = all_df.loc[mark0, 'pca_x'], y = all_df.loc[mark0, 'pca_y'], mode='markers', name='0'))
# fig.add_trace(go.Scatter(x = all_df.loc[mark1, 'pca_x'], y = all_df.loc[mark1, 'pca_y'], mode='markers', name='1'))
# fig.add_trace(go.Scatter(x = all_df.loc[mark2, 'pca_x'], y = all_df.loc[mark2, 'pca_y'], mode='markers', name='2'))
# fig.add_trace(go.Scatter(x = all_df.loc[mark3, 'pca_x'], y = all_df.loc[mark3, 'pca_y'], mode='markers', name='3'))

In [44]:
all_df.reset_index(inplace=True, drop=True)

In [45]:
all_df.describe()

h_1           h_2           h_3           h_4           h_5  \
count  16000.000000  16000.000000  16000.000000  16000.000000  16000.000000   
mean       0.006443     29.616901      0.710119     51.887397      0.334333   
std        0.006769     27.952025      0.127883     44.491989      0.126889   
min        0.000412      0.509750      0.345329      3.803608      0.091838   
25%        0.002946     13.743019      0.605065     25.741743      0.258151   
50%        0.004300     21.034090      0.711826     38.809960      0.319617   
75%        0.007084     33.751165      0.817389     61.366431      0.391337   
max        0.089301    232.101446      0.981387    441.068101      0.846592   

                h_6           h_7           h_8           h_9          h_10  \
count  16000.000000  16000.000000  16000.000000  16000.000000  16000.000000   
mean     254.033966    177.932687      5.530751      8.513955      0.000970   
std       24.214158    159.505947      0.481478      1.092801      0.000541   
min      142.024191     12.694610      3.657954      4.732163      0.000201   
25%      237.837282     86.316135      5.214266      7.915793      0.000646   
50%      252.503228    132.286067      5.508192      8.594567      0.000866   
75%      270.488738    212.522940      5.846538      9.097767      0.001090   
max      304.458172   1638.944759      7.043865     11.636365      0.004853   

       ...         lbp_6         lbp_7         lbp_8         lbp_9  \
count  ...  16000.000000  16000.000000  16000.000000  16000.000000   
mean   ...      0.042372      0.040865      0.113624      0.126322   
std    ...      0.021452      0.016770      0.008093      0.015495   
min    ...      0.012000      0.007111      0.082044      0.085022   
25%    ...      0.028800      0.029278      0.108800      0.115600   
50%    ...      0.039822      0.041422      0.113022      0.123556   
75%    ...      0.050089      0.050889      0.118356      0.133478   
max    ...      0.187200      0.106933      0.159644      0.317511   

             lbp_10        bright        target        kmeans         pca_x  \
count  16000.000000  16000.000000  16000.000000  16000.000000  16000.000000   
mean       0.367936    135.886063      1.500000      0.592250      0.000000   
std        0.039202     12.168965      1.118069      0.898006    166.461477   
min        0.126800     77.000000      0.000000      0.000000   -173.815529   
25%        0.351289    128.000000      0.750000      0.000000    -96.190346   
50%        0.373111    135.000000      1.500000      0.000000    -48.467402   
75%        0.396667    144.000000      2.250000      1.000000     36.187652   
max        0.424400    161.000000      3.000000      3.000000   1510.641127   

              pca_y  
count  16000.000000  
mean       0.000000  
std       28.032600  
min      -92.315842  
25%      -22.803077  
50%        1.480054  
75%       20.358971  
max      121.583507  

[8 rows x 28 columns]

In [46]:
all_df.loc[mark0].index

Int64Index([    6,     7,    14,    15,    22,    23,    30,    31,    38,
               39,
            ...
            15973, 15976, 15979, 15981, 15984, 15987, 15989, 15992, 15995,
            15997],
           dtype='int64', length=9674)